src/plugins/intel_cpu/src/transformations/snippets/tpp/pass/lowered/set_tpp_leading_dim.cpp

```diff
@@ -0,0 +1,163 @@
+                return false;
+        }
+        return true;
+    };
+    bool has_buffer = false;
+    const auto& orig_loop_ids = port.get_expr()->get_loop_ids();
+    for (const auto& p : port.get_connected_ports()) {
+        const auto& connected_expr = p.get_expr();
+        if (ov::is_type<snippets::op::Buffer>(connected_expr->get_node()) &&
+            accepted_loops(orig_loop_ids, connected_expr->get_loop_ids())) {
+            OPENVINO_ASSERT(!has_buffer, "Only one Buffer can be connected to a TPP op");
+            has_buffer = true;
+        }
+    }
+    return has_buffer;
+}
+
+size_t get_leading_dim(ExpressionPort port, const snippets::lowered::LinearIR::LoopManagerPtr& loop_mngr) {
+    const auto& port_desc = port.get_descriptor_ptr();
+    auto layout = port_desc->get_layout();
+    auto shape = port_desc->get_shape();
+    auto subtensor = port_desc->get_subtensor();
+    // Some expressions (e.g. ReduceMax/ReduceSum) allow for FULL_DIM values in subtensor.
+    // Here we should replace them with actual dim values before calculating strides & offsets.
+    bool full_dim_substituted = false;
+    for (size_t i = 1; i <= subtensor.size(); i++) {
+        const auto idx = subtensor.size() - i;
+        if (subtensor[idx] == snippets::lowered::PortDescriptor::ServiceDimensions::FULL_DIM) {
+            // the reason that we don't support FULL_DIM substitution for an arbitrary layout is that
+            // the layout and subtersor can (and usually do) have different ranks
+            full_dim_substituted = true;
+            subtensor[idx] = shape[shape.size() - i];
+        }
+    }
+    OPENVINO_ASSERT(!full_dim_substituted || is_planar_layout(layout),
+                    "Only planar layouts are supported for FULL_DIM substitution");
+
+    if (has_directly_connected_buffer(port, loop_mngr)) {
+        shape = port_desc->get_subtensor();
+        OPENVINO_ASSERT(is_planar_layout(layout), "Only planar layouts are supported for Buffers");
+        const auto rank_diff = static_cast<int64_t>(layout.size()) - static_cast<int64_t>(shape.size());
+        if (rank_diff > 0)
+            layout.erase(layout.end() - rank_diff, layout.end());
+    }
+
+    OPENVINO_ASSERT(layout.empty() || (layout.back() == layout.size() - 1 && layout.size() == shape.size()),
+            "get_leading_dim detected invalid layout values: check shape + layout combination");
+    const auto dim = [&]() -> size_t {
+            switch (port.get_type()) {
+            // Input shape is original, so we need to correctly read this data by order
+            // Example:
+            //      Original shape (shape) = [1, 49, 2, 23]
+            //      Layout (transpose order) = [2, 0, 1, 3]
+            //      Transposed shape = [2, 1, 49, 23]
+            //      The leading dimension is equal to stride of shape[layout[3]] = 2 x 23
+            case ExpressionPort::Type::Input :
+                return layout[layout.size() - 2]; // `1` in example
```

Comment:
We can but this one was left intentionally, so both `Input` and `Output` cases use the same `layout.size() - 2` index.
I believe it is easier to understand the code this way 🙂

Reply:
Anyway, Both `Input` and `Output` cases use the same `1` index in `utils::get_input_dim_idx` and `utils::get_ouput_dim_idx` as well... Because when I see `std::distance(,,std::find,... .size() - 2))` I should spend some time to understand what it means 😄 
Although it's just a suggestion 😃

Reply:
I see your point now! So all this `std::distance(...)` stuff is in the `get_output_dims` helper. Then we can reuse it for sure. Thanks!